### PR TITLE
drive train code

### PR DESCRIPTION
**Pull request by HudsonHurtig**
plz review and ask questions! Note that there are two different shifting conditionals for two different types of "shifting"

### CODE REVIEW

**Comment by HudsonHurtig** [comment]
also note that the Changes in the third file are the new vendor repositories I had to add for the  NEOs

**Comment by HudsonHurtig** [comment]
the Compressor i was wanting to talk to you guys on exactly how(where) we will charge the compressor but the import is in Robot for now.

File: src/main/java/frc/robot/drive.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    public static int PCM_CAN_ID = 19;
+    public static boolean driveMechToggle;
+    public static void initDrive(){
+        L1 = new CANSparkMax(10, MotorType.kBrushless);
+        L2 = new CANSparkMax(11, MotorType.kBrushless);
+        R1 = new CANSparkMax(12, MotorType.kBrushless);
+        R2 = new CANSparkMax(13, MotorType.kBrushless);
+        L2.follow(L1);
+        R2.follow(R1);
+        driveController = new XboxController(1);
```
Comment (cyrusduong) [inline src/main/java/frc/robot/drive.java:32]: Consider the following:
Can we accomplish the duties of the drive class without actually having full control over the controller?
Can we work with just perhaps just the axis values (a range between -1 and 1)?
Do we *need* to initialize or have the controller here/in this class?

I want you to consider the possibility that the "drive controller" can be used to do other actions than just in the Drive class and hesitate about the decision to initialize it here.

File: src/main/java/frc/robot/drive.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+package frc.robot;
+
+
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.CANSparkMaxLowLevel.MotorType;
+import edu.wpi.first.wpilibj.DoubleSolenoid;
+import edu.wpi.first.wpilibj.DoubleSolenoid.Value;
+
+
+import edu.wpi.first.wpilibj.XboxController;
```
Comment (cyrusduong) [inline src/main/java/frc/robot/drive.java]: Nit: Normally we just have a single newline between the package name and the imports. The rest of the newlines can be cleaned up.

File: src/main/java/frc/robot/Robot.java [hunk 3]
```diff
@@ -86,6 +89,11 @@ public void autonomousPeriodic() {
    */
   @Override
   public void teleopPeriodic() {
+    double xAxisDemand = drive.driveController.getX(Hand.kRight) * -.3; //get joystick values
+    double yAxisDemand = drive.driveController.getY(Hand.kLeft) * -1;
+    boolean turnStickButton = drive.driveController.getStickButton(Hand.kRight); //this is the button taht is pressed if you push down on the joysticks, or in this case the right joystick
+    boolean toggleY = drive.driveController.getYButton(); //gets raw y button value, toggle stuff is all done in the arcade drive function
```
Comment (cyrusduong) [inline src/main/java/frc/robot/Robot.java]: Relevant to comment in Drive class. Do we want one of the controllers to exist in the drive class? Or should it only pass in what is necessary (joystick values and fucntion calls to shift) to Drive class to accomplish its function?

File: src/main/java/frc/robot/drive.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * Add your docs here.
+ */
+public class drive {
+    public static CANSparkMax L1, L2, R1, R2;
+    public static XboxController driveController;
+    public static DoubleSolenoid rDriveShift, lDriveShift;
+    public static int PCM_CAN_ID = 19;
+    public static boolean driveMechToggle;
+    public static void initDrive(){
```
Comment (cyrusduong) [inline src/main/java/frc/robot/drive.java]: Nit: Can we put a newline here clean separation between variable definition and the start of method definitions.

File: src/main/java/frc/robot/drive.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        R1 = new CANSparkMax(12, MotorType.kBrushless);
+        R2 = new CANSparkMax(13, MotorType.kBrushless);
+        L2.follow(L1);
+        R2.follow(R1);
+        driveController = new XboxController(1);
+        rDriveShift = new DoubleSolenoid(PCM_CAN_ID, 0, 7); //these values are subject to change if wiring changes
+        lDriveShift = new DoubleSolenoid(PCM_CAN_ID, 1, 6); //^^
+        driveMechToggle = false;
+    }
+    public static void arcadeDrive(double forwardPower, double turnPower, boolean powerBoost, boolean mechanicalBoost){
```
Comment (cyrusduong) [inline src/main/java/frc/robot/drive.java]: Nit: Can we put a new line here between method definitions.

File: src/main/java/frc/robot/drive.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        /*
+        this arcade drive function has two shiting modes one shifts a power scaler and is then applied to only forward power,
+         but could easily be switched to both forward power and turn power based on driver preference. the second shifting mode is the actual gear shift.
+         the final part is the math for arcade drive
+         */
+        if(powerBoost){  //if power boost true, give raw input to the motor output: which to default is half input the power and so by pressing and holding you double the output power to the motors from whats normal
+            forwardPower = forwardPower *  1.0;
+        } else {
+            forwardPower = forwardPower *  0.5; //looky
+        }
```
Comment (cyrusduong) [inline src/main/java/frc/robot/drive.java:49]: I like what is happening here. However, do we want this responsibility to fall under the Drive class?

I prefer to keep the units of work/functions as small as possible (Drive functions) and leave the consumer of this class (Robot) to consume it however they wish. So if they wanted to halve the forward power/turning power, they should do so before it is passed in via the `arcadeDrive` function.

I'm ok with however we want to do it and ultimately it's up to the team to decide, but the simpler your code can be, typically the easier it is to test/troubleshoot.

File: src/main/java/frc/robot/drive.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+        if(mechanicalBoost){driveMechToggle = !driveMechToggle;} //this toggles a boolean. that way, to switch to high speeds(mechanically) you just press the button once to switch the gear ratios
+
+        if(mechanicalBoost){  //shift gears to lower torque:higher RPM ratio
```
Comment (cyrusduong) [inline src/main/java/frc/robot/drive.java]: I think there are a few logical errors here.

1. Line 52 - There is a toggle that changes when `mechanicalBoost` is `true`. What happens when `mechanicalBoost` stays true for multiple loops? (i.e. - The button that returns this boolean is held down for a period of time?). 

2. Line 54 - The same `mechanicalBoost` is used to set the shifters. Not only incurring the error/problem I describe in (1), but also leaves the variable `driveMechToggle` unused. Should `driveMechToggle` be used here instead?

Lastly Nit: The variable name `driveMechToggle` is not very descriptive about what state it is holding or what it is used for. Would something like `shiftersIsHighGear` or `shiftersToggleLowGear` more appropriate?

File: src/main/java/frc/robot/Robot.java [hunk 8]
```diff
@@ -10,7 +10,8 @@
 import edu.wpi.first.wpilibj.TimedRobot;
 import edu.wpi.first.wpilibj.smartdashboard.SendableChooser;
 import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
-
+import edu.wpi.first.wpilibj.GenericHID.Hand;
+import edu.wpi.first.wpilibj.Compressor;
```
Comment (cyrusduong) [inline src/main/java/frc/robot/Robot.java]: Is this unused? For now, we might want to consider removing it unless we are going to do the work in this branch/PR to make it use of it.

**Review by cyrusduong — CHANGES_REQUESTED** [review]
Overall a very good pull request. There are some nits I have about it but only a few things that are of real concern. Address these and it should be good.

From there we can make more improvements to it like maybe traction-control/auto-shifting and more. For now small steps. 

We can discuss this more in tonight's meeting if you are confused about anything. We'll work through it together.

**Comment by cyrusduong** [comment]
I'll also see if I can get a linter/analysis tool to run that'll help with all my `nit: comments` so we can focus on the 'logical' code and decisions made in the code.

Important to keep the `code quality` consistent across the team no matter who is writing it. 😃

File: src/main/java/frc/robot/drive.java (hunk 1 above)
Comment (HudsonHurtig) [reply src/main/java/frc/robot/drive.java:32]: discussed in person resolving later

File: src/main/java/frc/robot/Robot.java (hunk 3 above)
Comment (HudsonHurtig) [reply src/main/java/frc/robot/Robot.java]: discussed in person

**Review by cyrusduong — APPROVED** [review]

**Review by rahidas04 — APPROVED** [review]

**Pull request merged**